### DISCUSSION TITLE
Can’t sign a transaction

### DISCUSSION THREAD

**Discussion by itsallmememe** [WalletConnect/walletconnect-monorepo#2537] [Developer Support]
I am trying to get WC to sign a transaction, but failing. Rainbow opens to perform the connect process, my app then opens, but the sign code does not re-open the wallet. If I manualy switch to Rainbow it does not show the sign popup.

- I’ve tried adding a delay in before I do the ` Sign.instance.request` (thought it might be something to do with being in the background)
- I’ve tried adding in my “open wallet” code which takes the uri and universal link code to open the wallet again
- I’ve tried creating a new uri before `Sign.instance.request` and opening the app with that uri

but none of it works

Any ideas what I can do?
- is it perhaps that I’m not sending in enough information into the `Request.params`?

Connect code
```
            let methods: Set<String> = ["eth_sendTransaction", "personal_sign", "eth_signTypedData"]

            var blockchains: Set<Blockchain>

            if Bundle.main.isDevelopmentBuild {
                blockchains = Set([.MainNet, .Sepolia, .Goerli])
            } else {
                blockchains = Set([.MainNet])
            }

            chainId = .MainNet

            let namespaces: [String: ProposalNamespace] = ["eip155": ProposalNamespace(chains: blockchains, methods: methods, events: [])]

            do {
                try await Sign.instance.connect(requiredNamespaces: namespaces, topic: uri.topic)
            } catch {
                notificationCenter.post(name: .walletFailedtoConnect, object: self, userInfo: [:])
                try await createUri()
            }

            Task { @MainActor in
                let url = URL(string: "https://\(wallet.universalLinkHost)/wc?uri=\(uri.deeplinkUri)")
                UIApplication.shared.open(url!)
            }
```

and the signing code
``` 
Sign.instance
            .sessionSettlePublisher
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [unowned self] session in
                print("sessionSettlePublisher", session)

                Task { @MainActor in
                    let method = "personal_sign"
                    let requestParams = AnyCodable([publishSignModel.from, publishSignModel.to, publishSignModel.data])
                    let request = Request(topic: session.topic,
                                          method: method,
                                          params: requestParams,
                                          chainId: chainId!)
                    try await Sign.instance.request(params: request)
                }
            })
            .store(in: &subscriptions)
```

**Comment by finessevanes** [comment]
Hey @itsallmememe could you share what SDK and version you're using?

**Reply by pqseags** [reply]
Hey @finessevanes , I'm working with @itsallmememe on this. We're using the Swift iOS SDK, WalletConnect v1.6.6

**Reply by finessevanes** [reply]
Thanks - pinging our Swift team for support.

**Comment by pqseags** [comment]
I made some progress here - with these changes I'm able to get a notification from Rainbow to sign a transaction, which when clicked on the transaction populates in Rainbow and is able to be signed. Would still appreciate some guidance on best practices or additional documentation on how to encode the data to send a transaction, and if there is a way to open Rainbow automatically with the transaction populated.

* We actually want the method `eth_sendTransaction`
* It looks like we can format the request params in the following way, although I'm not sure which fields are required/recommended (like gas, gasPrice, etc)

```
let method = "eth_sendTransaction"
struct EthSendTransaction: Codable, Equatable {
    let from: String
    let to: String
    let data: String
}
let requestParams = AnyCodable([EthSendTransaction(from: publishSignModel.from, to: publishSignModel.to, data: publishSignModel.data)])
```

**Reply by alexander-lsvk** [reply]
Hey @itsallmememe @pqseags,
[Here](https://github.com/WalletConnect/WalletConnectSwiftV2/blob/49b2975088808bec0217110eb977be17711d4a0f/Example/DApp/Sign/AccountRequest/AccountRequestViewController.swift#L99) you can find an example of `eth_sendTransaction` payload.
@itsallmememe could you please provide more context? Were you able to set up a session with Rainbow Wallet?

**Comment by itsallmememe** [comment]
Hi @alexander-lsvk - thanks for the reply

We have everything working now - auth, connect and sign! 

I have two small questions tho. When I use a uri to connect and open the wallet, can I reuse that websocket connection between signings and app sessions? Or do I need to recreate every time?

If I can reuse between app sessions, how to I preload this into the Sign SDK?

Thanks!
